zebra-chain/src/transaction.rs

```diff
@@ -185,4 +185,27 @@ impl Transaction {
     ) -> blake2b_simd::Hash {
         sighash::SigHasher::new(self, hash_type, network_upgrade, input).sighash()
     }
+
+    /// Access the joinsplits of this transaction, regardless of version.
+    // Todo: signature haves to be like:
+    // pub fn joinsplits<P: ZkSnarkProof>(&self) -> Option<JoinSplitData<P>> {
```

Comment:
This should use the `ZkSnarkProof` market trait in order to return `Bctv14Proof` or `Groth16Proof` however i don't know how to use it properly. Specifically how to return(just returning `*joinsplit_data` will not work).

For my specific purposes(check if there are joinsplits inside a transaction) i modified the function but i don't think this is what we want.

Reply:
@yaahc pinging you for this one. Thanks!

Reply:
I thikn the point of the marker trait is that we cannot ever combine these types into one type, since they're encoded differently. The definition as is describes a function where the caller provides a ZkSnarkProof type, and you return a `JointSplitData<P>` where P is the exact type they requested. What we want is to return the exact type we actually have depending on which transaction variant self is, rather than having the caller select the expected type for us.

I think the precise signature you'd want to define is something like

```rust
fn joinsplits(&self) -> Option<Either<JoinSplitData<Bctv14Proof>, JoinSplitData<Groth16Proof>>>;
```

Reply:
Thanks @yaahc and please excure my ignorance about the marker trait. Your signature works great, i implemented it at https://github.com/ZcashFoundation/zebra/pull/1190/commits/ce00e72aca72f1101f81af69007bda9b2479472a